Shell/AnswerExtractor.hpp

```diff
@@ -61,22 +69,79 @@ class AnswerLiteralManager : public AnswerExtractor
   void addAnswerLiterals(Problem& prb);
   bool addAnswerLiterals(UnitList*& units);
 
-  bool isAnswerLiteral(Literal* lit);
+  virtual void onNewClause(Clause* cl);
 
-  void onNewClause(Clause* cl);
+  virtual Clause* recordAnswerAndReduce(Clause* cl) { return nullptr; };
 
-private:
+protected:
+  Clause* getRefutation(Clause* answer);
   Literal* getAnswerLiteral(VList* vars,Formula* f);
+
+private:
   Unit* tryAddingAnswerLiteral(Unit* unit);
 
+  virtual Formula* tryGetQuantifiedFormulaForAnswerLiteral(Unit* unit);
+
+  virtual Unit* createUnitFromConjunctionWithAnswerLiteral(Formula* junction, VList* existsVars, Unit* originalUnit);
+
   Clause* getResolverClause(unsigned pred);
-  Clause* getRefutation(Clause* answer);
 
   RCClauseStack _answers;
 
   DHMap<unsigned, Clause*> _resolverClauses;
 };
 
+class SynthesisManager : public AnswerLiteralManager
+{
+public:
+  static SynthesisManager* getInstance();
+
+  virtual bool tryGetAnswer(Clause* refutation, Stack<TermList>& answer) override;
+
+  virtual void onNewClause(Clause* cl) override;
+
+  virtual Clause* recordAnswerAndReduce(Clause* cl) override;
+
+  static unsigned getITEFunctionSymbol(TermList sort) {
```

Comment:
This isn't polymorphism-safe, and introduces one symbol per sort in the many-sorted case. Could this be a single polymorphic symbol `$ite: !>[X: $tType]: ($o * X * X) > X` instead?

Reply:
After discussion: this might be possible, but polymorphism and synthesis needs further investigation in a separate PR anyway. Keep it as-is for now.